torch_xla/core/xla_model.py

```diff
@@ -1299,6 +1299,19 @@ def convert_fn(tensors):
     shardings = None
     if input_sharding:
       shardings = [input_sharding.xla_spec(t) for t in tensors]
+    if input_sharding and input_sharding.minibatch:
+      # when minibatch is configured we must make sure batch dimension of
+      # the tensor is divisible by the data mesh dimension.
+      data_mesh_dim = input_sharding.mesh.mesh_shape[0]
+      for tensor, sharding in zip(tensors, shardings):
+        # assume batch dimension is 0
+        if sharding and tensor.dim() > 0 and (tensor.size()[0] %
```

Comment:
Trying to understand the issue a bit more - `minibatch` means each host is loading data for its local devices, and the result will be distributed across the local devices and viewed as a global tensor.

So if each host loads 3 examples and has 4 devices, we'll end up with host0 having `[1,2,3,0]` after padding its local example and host1 having `[4,5,6,0]` after padding its own.

Then the final global tensor has data `[1,2,3,0,4,5]` (after concat/unpadding) instead of the desired `[1,2,3,4,5,6]`?

If that's the case, then we should compare the local batch size to the local device count instead of the local shard to the mesh shape?

(I don't recall the full semantics of the API here, but I believe we take in a tensor of the local batch and output a tensor representing the global batch when `minibatch=True`)

Reply:
oh you right.. but the issue is that we might have a 2d mesh of `(4,4,None,None)`. I was trying to fix that issue here.. Now I think about it, if that's the case what should I do...

Reply:
easier solution is we alwasy assume 1d mesh for the data loading, which I believe is almost always the case.

Reply:
I think @khatwanimohit tried to explicitly disallow higher-dim meshes (ie we only support all devices along a single axis): https://github.com/pytorch/xla/blob/365841da4b3497a832b54045eb0c3ac39c4a1e47/torch_xla/csrc/init_python_bindings.cpp#L1299-L1301

Reply:
> easier solution is we alwasy assume 1d mesh for the data loading, which I believe is almost always the case.

Yeah, IIRC that's also the approach taken in MaxText - load everything with pure batch-dim sharding, and let the compiler shuffle the data around to match the model sharding. Mohit is the expert here though!

Reply:
sg, will update